ports/esp32/esp32_pcnt.c

```diff
@@ -0,0 +1,450 @@
+    if (n_pos_args == 1) {
+        unit = mp_obj_get_int(args[0]);
+        if (unit < 0 || unit >= PCNT_UNIT_MAX) {
+            mp_raise_ValueError(MP_ERROR_TEXT("invalid id"));
+        }
+    }
+
+    esp32_pcnt_obj_t *self = MP_STATE_PORT(esp32_pcnt_obj_head);
+    mp_uint_t allocated = 0;
+    while (self != NULL) {
+        if (self->unit == unit) {
+            break;
+        }
+        allocated |= 1 << self->unit;
+        self = self->next;
+    }
+    if (self == NULL) {
+        if (unit == PCNT_UNIT_MAX) {
+            for (unit = 0; unit < PCNT_UNIT_MAX; unit++) {
+                if (!(allocated & 1)) {
+                    break;
+                }
+                allocated >>= 1;
+            }
+            if (unit == PCNT_UNIT_MAX) {
+                mp_raise_TypeError(MP_ERROR_TEXT("no unused PCNT unit"));
+            }
+        }
+        self = m_new_obj(esp32_pcnt_obj_t);
+        self->base.type = &esp32_pcnt_type;
+        self->unit = unit;
+        self->next = MP_STATE_PORT(esp32_pcnt_obj_head);
+        MP_STATE_PORT(esp32_pcnt_obj_head) = self;
+        esp32_pcnt_deinit_helper(self);
+    }
+
+    mp_map_t kw_args;
+    mp_map_init_fixed_table(&kw_args, n_kw_args, args + n_pos_args);
+    esp32_pcnt_init_helper(self, 0, args + n_pos_args, &kw_args);
+
+    if (!pcnt_isr_service_installed) {
+        check_esp_err(pcnt_isr_service_install(ESP_INTR_FLAG_IRAM));
+        pcnt_isr_service_installed = true;
+    }
+    check_esp_err(pcnt_intr_enable(self->unit));
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC void esp32_pcnt_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    esp32_pcnt_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "PCNT(%u)", self->unit);
+}
+
+STATIC mp_obj_t esp32_pcnt_init(mp_uint_t n_pos_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    esp32_pcnt_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+    esp32_pcnt_init_helper(self, n_pos_args - 1, pos_args + 1, kw_args);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(esp32_pcnt_init_obj, 1, esp32_pcnt_init);
+
+STATIC mp_obj_t esp32_pcnt_deinit(mp_obj_t self_in) {
+    esp32_pcnt_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    esp32_pcnt_deinit_helper(self);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(esp32_pcnt_deinit_obj, esp32_pcnt_deinit);
+
+STATIC mp_obj_t esp32_pcnt_value(size_t n_args, const mp_obj_t *pos_args) {
+    esp32_pcnt_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+    int16_t value;
+    if (n_args == 2) {
+        value = mp_obj_get_int(pos_args[1]);
+        if (value != 0) {
+            mp_raise_ValueError(MP_ERROR_TEXT("can only set value to 0"));
+        }
+    }
+    while (true) {
+        check_esp_err(pcnt_get_counter_value(self->unit, &value));
+        if (self->pending && self->handler != MP_OBJ_NULL) {
+            esp32_pcnt_run_irq_handlers(NULL);
+        } else {
+            break;
+        }
+    }
```

Comment:
I'm not sure, looks like a deadlock if value() is called from an isr.
I'm not sure, looks like a deadlock if value() is called from an isr.
I'm not sure, looks like a deadlock if value() is called from an isr.
I'm not sure, looks like a deadlock if value() is called from an isr.
I'm not sure, looks like a deadlock if value() is called from an isr.
```
def cb():
     print(encoder.value())
```

Reply:
The pending flag is cleared before entering the callback, so the callback will only re-enter itself via a call to `.value()` if another interrupt occurs before the call to `.value()`. Even in that case, it should only deadlock if the callback can't be re-entered for some reason (i.e., it acquires some non-reentrant lock).

`.value()` itself is safe to re-enter and I've tested this exact code path. If you suffered some kind of interrupt storm (perhaps by misconfiguring the limits on very fast moving counter) then you might run out of stack, but it won't deadlock.

There's an argument to not allowing `.value()` to trigger callbacks if it is called from within a callback, but detecting that would require additional careful use of state. At the moment, I'm more inclined to re-write the interrupt stuff completely to support both soft and hard callbacks, but that will have to wait until I have more time…